From 039a5efbb23c92c82956ebbbf09c4a5b2c0b1c0a Mon Sep 17 00:00:00 2001
From: "Pierre-Loup A. Griffais" <pgriffais@valvesoftware.com>
Date: Wed, 8 Nov 2023 19:45:52 -0800
Subject: [PATCH 18/39] drm/amdgpu/display: fix backlight brightness
 overflowing

500k uNits * 65k max brightness range overflows in the conversion code.
Use 64 bit dividend to avoid the overflow

Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
index cb64c80f9..c25f862b4 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
@@ -4850,7 +4850,7 @@ static u32 convert_brightness_from_user(const struct amdgpu_dm_backlight_caps *c
 	}
 
 	// Rescale 0..255 to min..max
-	return min + DIV_ROUND_CLOSEST((max - min) * brightness,
+	return min + DIV_ROUND_CLOSEST_ULL((u64)(max - min) * brightness,
 				       AMDGPU_MAX_BL_LEVEL);
 }
 
@@ -4865,7 +4865,7 @@ static u32 convert_brightness_to_user(const struct amdgpu_dm_backlight_caps *cap
 	if (brightness < min)
 		return 0;
 	// Rescale min..max to 0..255
-	return DIV_ROUND_CLOSEST(AMDGPU_MAX_BL_LEVEL * (brightness - min),
+	return DIV_ROUND_CLOSEST_ULL((u64)AMDGPU_MAX_BL_LEVEL * (brightness - min),
 				 max - min);
 }
 
-- 
2.49.0

